File: lib/tree.rb
class TreeNode
  attr_reader :key, :value
  attr_accessor :left, :right
  
  def initialize(key, val)
    @key = key
    @value = val
    @left = nil
    @right = nil
  end
end

class Tree
  attr_reader :root
  def initialize
    @root = nil
  end
  
  # Time Complexity: O(n) worst case
  # Space Complexity: O(1)
  def add(key, value)
    @root = add_helper(@root, key, value)
  end

  def add_helper(current_node, key, value)
    if current_node.nil?
      return TreeNode.new(key, value)
    end 
    
    if key < current_node.key
      current_node.left = add_helper(current_node.left, key, value)
    else
      current_node.right = add_helper(current_node.right, key, value)
    end 
    return current_node
  end 
  
  # Time Complexity: O(log(n) best case, O(n) worst case
  # Space Complexity: O(1)
  def find(key)
    return find_helper(@root, key)
  end
  
  def find_helper(current_node, key)
    if current_node.nil?
      return nil 
    end 
    
    if key > current_node.key 
      current_node.right = find_helper(current_node.right, key)
    elsif key < current_node.key
      current_node.left = find_helper(current_node.left, key)
    else
      return current_node.value
    end
  end

  # Time Complexity: O(n)
  # Space Complexity:  O(n)
  def inorder
    arr = []
    if @root.nil?
      return arr
    end 
    current_node = @root
    return inorder_helper(current_node, arr)
  end
  
  def inorder_helper(current_node, arr)
    if current_node.nil?
      return arr
    end 
    inorder_helper(current_node.left, arr)
    arr << {key: current_node.key, value: current_node.value}
    inorder_helper(current_node.right, arr)
    return arr
  end
  
  # Time Complexity: O(n)
  # Space Complexity: O(n)
  def preorder
    arr = []
    if @root.nil?
      return arr
    end 
    current_node = @root
    return preorder_helper(current_node, arr)
  end
  
  def preorder_helper(current_node, arr)
    if current_node.nil?
      return arr
    end 
    arr << { key: current_node.key, value: current_node.value }
    preorder_helper(current_node.left, arr)
    preorder_helper(current_node.right, arr)
    return arr
  end
  
  # Time Complexity: O(n)
  # Space Complexity: O(n)
  def postorder
    arr = []
    if @root.nil?
      return arr
    end 
    current_node = @root
    return postorder_helper(current_node, arr)
  end
  
  def postorder_helper(current_node, arr)
    if current_node.nil?
      return arr
    end 
    postorder_helper(current_node.left, arr)
    postorder_helper(current_node.right, arr)
    arr << { key: current_node.key, value: current_node.value }
    return arr
  end
  
  # Time Complexity: O(n)
  # Space Complexity: O(n) - stack
  def height(current_node = @root)
    if current_node.nil?
      return 0
    end
    
    r = height(current_node.right)
    l = height(current_node.left)
    
    return [r, l].max + 1
  end

  # did not decide to do optionals
end

File: binary_search_tree/tree.py
class TreeNode:
    def __init__(self, key, val = None):
        if val == None:
            val = key

        self.key = key
        self.value = val
        self.left = None
        self.right = None
        


class Tree:
    def __init__(self):
        self.root = None

    # Time Complexity: 0(log(n)
    # Space Complexity: 0(1)
    def add(self, key, value = None):
        added = TreeNode(key, value)
        if not self.root:
            self.root = added

        parent = self.root
        while True:
            if added.key > parent.key:
                if not parent.right:
                    parent.right = added
                    return added
                else:
                    parent = parent.right
            elif added.key <= parent.key:
                if not parent.left:
                    parent.left = added
                    return added
                else:
                    parent = parent.left
            else:
                return "something went wrong"


    # Time Complexity: 0(log(n))
    # Space Complexity: 0(1)
    def find(self, key):
        current = self.root
        while current:
            if key == current.key:
                return current.value
            elif key <= current.key:
                current = current.left
            elif key > current.key:
                current = current.right

        return None

    def inorder_helper(self, current, nodes = None):
        if nodes is None:
            nodes = []
        if current:
            self.inorder_helper(current.left, nodes)
            nodes.append({
                "key": current.key,
                "value": current.value
            })
            self.inorder_helper(current.right, nodes)
        return nodes

    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def inorder(self):
        return self.inorder_helper(self.root)

    def preorder_helper(self, current, nodes = None):
        if nodes is None:
            nodes = []
        if current:
            nodes.append({
                "key": current.key,
                "value": current.value
            })
            self.preorder_helper(current.left, nodes)
            self.preorder_helper(current.right, nodes)
        return nodes

    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def preorder(self):
        return self.preorder_helper(self.root)

    def postorder_helper(self, current, nodes = None):
        if nodes is None:
            nodes = []
        if current:
            self.postorder_helper(current.left, nodes)
            self.preorder_helper(current.right, nodes)
            nodes.append({
                "key": current.key,
                "value": current.value
            })
        return nodes

    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def postorder(self):
        return self.postorder_helper(self.root)

    def height_helper(self, current, nodes = None):
        if nodes is None:
            nodes = []
        if current:
            self.height_helper(current.left, nodes)
            if current.left is None and current.right is None:
                nodes.append(current.key)
            self.height_helper(current.right, nodes)
        return nodes

    def get_count(self, key):
        current = self.root
        count = 0
        while current:
            count += 1
            if key == current.key:
                return count
            elif key > current.key:
                current = current.right
            elif key < current.key:
                current = current.left
        return None

    # Time Complexity: O(nlog(n))
    # Space Complexity: 0(m) -> m == # of nodes
    def height(self):
        ends = self.height_helper(self.root)
        max = 0
        for end in ends:
            height = self.get_count(end)
            if height > max:
                max = height
        return max

#   # Optional Method
#   # Time Complexity: 
#   # Space Complexity: 
    def bfs(self):
        pass

#   # Useful for printing
    def to_s(self):
        return f"{self.inorder()}"
